[Benjamin Soh] iP

bensohh opened the pull request:
## DukePro

> "The Web as I envisaged it, we have not seen it yet. The future is still so much bigger than the past." - Tim Berners-Lee, Inventor of the World Wide Web [(source)](https://www.brainyquote.com/quotes/tim_bernerslee_409503)

DukePro frees your mind of having to remember things you need to do. It's

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/bensohh/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE!**

### Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
##
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

bensohh committed:
Completed Level-1 Task

bensohh committed:
Completed task Level-2

bensohh committed:
Completed task Level-3 and added documentation

bensohh committed:
Completed Level-4 and added documentation for more classes

bensohh committed:
Completed task A-TextUiTesting, removed the old method of adding tasks

bensohh committed:
Completed task Level-5

bensohh committed:
Completed task Level-6, removed irrelevant errors and added more errors handling for out of bounds number

bensohh committed:
Changed the way the chatbox work by changing the processes to enum and utilising switch

bensohh committed:
Update Java Classes to use more OOP

bensohh committed:
Fix Bug to save and display task status

bensohh committed:
Create branch for Save Feature in Chatbot

bensohh committed:
Create branch for Date Feature Level-8

bensohh committed:
Resolve conflict for merging

bensohh committed:
Create multiple packages for Java Classes

bensohh committed:
Merge remote-tracking branch 'upstream/add-gradle-support' to enable testing.

bensohh committed:
Setup Gradle to automate build, run and checkstyle

bensohh committed:
Create JUnit Tests for Ui, Parser and Storage

bensohh committed:
Edit comments to methods

bensohh committed:
Include JavaDocs for AddCommand class

bensohh committed:
Include JavaDocs for most Classes and Methods

bensohh committed:
Include Find Function for ChatBot

bensohh committed:
Merge branch 'branch-Level-9' to include Find Function

sharmaine1028 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+    private Ui ui;
+
+    /**
+     * Contructor to create an instance of Duke.
```
Spelling error: Constructor

sharmaine1028 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * Contructor to create an instance of Duke.
+     *
+     * @param filePath the String representation of the relative file path to store the data
```
@param filePath String representation of the relative file path to store the data

sharmaine1028 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * To kickstart the entire ChatBot.
```
Perhaps could write "Kickstarts the entire Chatbot" instead as first sentence should start in form of "Kickstarts...., Returns..., Sends..." etc.

sharmaine1028 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+     * To kickstart the entire ChatBot.
+     */
+    public void run() {
+        System.out.println(ui.showWelcome());
```
Why not make ui.showWelcome() function directly print out the greeting? I feel that this will be more in line with the Tell, Don't Ask principle.

sharmaine1028 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,120 @@
+     * @param taskType Integer value representing the type of task, 0-TODO Task,
+     *                 1-DEADLINE Task, 2-EVENT Task
+     * @param taskDetails String representation of the task description
+     * @return instance of AddCommand
```
I thought that constructor does not return anything in Java? No need for @return tag

sharmaine1028 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * Create the corresponding task based on the taskType in AddCommand and adds task
```
Creates

sharmaine1028 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,120 @@
+     * Create the corresponding task based on the taskType in AddCommand and adds task
+     * to the TaskList. The task list is then stored on a file under data/tasks.txt
+     *
+     * @param tasks a list to add and keep track of the tasks
```
Should be capital letter after parameter, aka @param tasks List to add and keep track of the tasks

sharmaine1028 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        //Exception: Throw an error when the second half after "delete" keyword is greater than task_list
+        if (taskNumber > tasks.size() || taskNumber < 1) {
```
Why is there taskNumber < 1 condition in this line when it is already handled above?

sharmaine1028 replied:
I'm assuming that taskNumber < 1 is the same check as isEmpty() check

bensohh replied:
Hi! Sorry for the confusion, the isEmpty() check is to check if the tasks (TaskList) is empty while taskNumber here refers to the input from the user, in the event when the tasks (TaskList) contains task and they enter delete 0, delete -1 or any number that is < 1 we will throw an error since there can only be taskNumber starting from 1 onwards. Thank you for taking your time to review.

sharmaine1028 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * A String representation of successfully executing the DeleteCommand.
+     *
+     * @return a String to notify the user that a task has been removed
```
Remember to capitalise after @return tag as well

sharmaine1028 commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,90 @@
+     * @return an instance of UnmarkCommand
+     */
+    public UnmarkCommand(String taskDetails) {
+        this.taskNumber = Integer.parseInt(taskDetails);
```
I'm not very sure if possible error that can be thrown by parseInt has been handled. For instance if user types in unmark hello when it should be unmark 1. I checked the Parser class and it does not seem to have handled this case. Perhaps can consider throwing some exception here or somewhere higher in the chain.

bensohh replied:
Thank you for catching something I missed out! I will be adding in a NumberFormat exception when parsing the inputs from the user to ensure that the command is correct.

sharmaine1028 reviewed (commented):
In general, looks pretty good. Import statements are in the right order for the files I saw, static imports first etc. Indentation and Egyptian style braces are also in place.

However, take note of the conventions for JavaDocs
1. First sentence as short summary of method, should be in present tense form "Adds..., Returns..., Sends..."
2. First letter for parameter description should be capitalised, E.g. @param zone Zone of position

Other than that, I think the Java coding standards were largely followed. Good job!

bensohh committed:
Fix the coding standards errors and suggestions

bensohh committed:
Add NumberFormatException and Debug text-ui-test

bensohh committed:
Create GUI for Duke

bensohh committed:
Use varargs for FindCommand

bensohh committed:
Extract the findTasks component from FindCommand Class to TaskList Class

The current implementation of execute method in FindCommand iterates through
the entire TaskList, compare each task's description with the keywords before
appending it to the resulting String to return.

The iteration of the TaskList and searching for the matching Tasks should be
encapsulated within the TaskList Class.

Let's abstract out the component for searching for matching Tasks via keywords
into a new method within the TaskList Class called findTasks.

This follows after the OOP principle and prevents weak SLAP.

bensohh committed:
Include Assertions in AddCommand Class and Storage Class

The else block in the execute method within AddCommand class assumes
the taskType to be 2 (which refers to Event tasks). The load method
in Storage class assumes that the directory and file are created correctly.

Validates the taskType before creating an Event task. To ensure that
the directory and file have been created before writing to or reading
from the file.

Let's,
* Include an assert statement to check for the correct taskType in the
else block of the execute method.
* Include assert statements after commands to create file and create
directory to check if the directory and file exist before proceeding.

Validate the assumptions made and terminate the program if there is
an error before proceeding on with the rest of the code.

bensohh committed:
Fix Code Quality based on Coding Standards

There exist old methods that are unused within the Ui Class after the
implementation of JavaFX.

Methods that are unreferenced should be removed according to the
Coding Standards and Practices.

Let's,
* Remove showLine method in Ui Class
* Remove displayWelcome method in Ui Class
* Remove readCommand method in Ui Class
* Remove commented out codes

bensohh committed:
Implement Stream for findTasks and compileTasks methods in TaskList Class

The two methods in TaskList Class iterate through the entire ArrayList(Task) to
check for keywords and add it to the compiled or matched list.

This can be done more efficiently via Streams using map and filter functions that
are under the Stream Class.

Let's convert the TaskList into a Stream<Task> and utilise the map and filter function
to perform the same task.

bensohh committed:
Merge pull request #1 from bensohh/branch-A-Assertions

Include Assertions in AddCommand Class and Storage Class

bensohh committed:
Merge pull request #2 from bensohh/branch-A-CodeQuality

Fix Code Quality based on Coding Standards

bensohh committed:
Merge branch 'master' into branch-A-Streams

bensohh committed:
Merge pull request #3 from bensohh/branch-A-Streams

Implement Stream for methods in TaskList Class

bensohh committed:
Set up Continuous Integration(CI)

No CI is implemented for the previous versions of the project.

It is a good for at least one member to implement it for the upcoming tP project.

Let's include the gradle.yml file under the directory .github/workflows

bensohh committed:
Implement C-Undo extension for the ChatBot

There are no methods to undo a command, the existing method to "undo"
requires user to manually utilise existing method to undo their previous
command.

Provide a way to undo the most recent command which allows
user to restore the ChatBot to the state before the command
was executed with a simple "undo" command.

Let's
* Create an UndoCommand Class to handle the logic of undo-ing a command
* Create a HelpCommand Class to handle the prompting of user with existing
commands and examples
* Include assert statements to test assumptions made
* Remove unnecessary methods like isExit()

bensohh committed:
Merge pull request #4 from bensohh/branch-C-Undo

Implement C-Undo extension for the ChatBot

bensohh committed:
Update README.md

bensohh committed:
Update README.md and HelpCommand

bensohh committed:
Update README.md and fix minor bugs

bensohh committed:
Add more test cases

bensohh committed:
Fix bugs and Coding Standards

bensohh committed:
Add Ui.png and Fix styling for GUI

bensohh committed:
Change background for GUI

bensohh committed:
Improve GUI and fix minor bugs
